src/encoded/types/dataset.py

```diff
@@ -213,6 +214,162 @@ def assembly(self, request, original_files, status):
     def hub(self, request):
         return request.resource_path(self, '@@hub', 'hub.txt')
 
+    @calculated_property(condition='analysis_objects', schema={
+        "title": "Default analysis",
+        "description": "One default analysis that should be checked first.",
+        "comment": "Do not submit. This field is calculated from files in this analysis.",
+        "type": "string",
+        "linkTo": "Analysis",
+        "notSubmittable": True
+    })
+    def default_analysis(self, request, status, analysis_objects):
```

Comment:
Not sure how much time you want to spend on this but I wonder if it makes sense to pull all of the ranking criteria out of every analysis, store it in a tuple, and then sort the tuples at the end. The tuples should sort one position at a time so I think you'll end up with the same answer (i.e. status rank is higher priority than pipeline_version rank).

It might look something like this (not sure if this code works at all):

```
 def default_analysis(self, request, status, analysis_objects):
        types = request.registry[TYPES]
        unreleased_status = ['in progress', 'submitted', 'deleted', 'replaced']
        # Easier to sort everything by max, so reverse order of all.
        status_order = list(reversed(types['analysis'].schema['properties']['status']['enum']))
        award_rfa_order = list(reversed(types['award'].schema['properties']['rfa']['enum']))
        assembly_order = list(reversed(types['file'].schema['properties']['assembly']['enum'] + ['mixed']))
        genome_annotation_order = list(reversed(types['file'].schema['properties']['genome_annotation']['enum'] + ['mixed']))
        analyses = (
            request.embed(
                at_id,
                '@@object_with_select_calculated_properties'
                '?field=@id'
                '&field=pipeline_labs'
                '&field=pipeline_award_rfas'
                '&field=assembly'
                '&field=genome_annotation',
            )
            for at_id in analyis_objects
        )
        if status not in unreleased_status:
            status_order = [
                s
                for s in status_order
                if s not in unreleased_status
            ]
            analyses = (
                analysis
                for analysis in analyses
                if  analysis['status'] in status_order
            )
        analysis_ranking = {}
        for analysis in analyses:
            status_rank = status_order.index(analysis['status'])
            # True ranked higher than False.
            lab_rank = '/labs/encode-processing-pipeline/' in analysis.get('pipeline_labs', [])
            # Get max index or zero if field doesn't exist. 
            pipeline_award_rfa_rank = max(
                award_rfa_order.index(rfa)
                for rfa in analysis.get('pipeline_award_rfa', []) + [0]
            )
            # Get max index or zero if field doesn't exist. 
            assembly_rank = max(
                assembly_order.index(assembly)
                for assembly in analysis.get('assembly', []) + [0]
            )
            # Get max index or zero if field doesn't exist. 
            genome_annotation_rank = max(
                genome_annotation_order.index(annotation)
                for annotation in analysis.get('genome_annotation', []) + [0]
            )
            # We reverse sort order at the end so later version numbers will rank higher.
            pipeline_version_rank = parse_version(analysis.get('pipeline_version', ''))
            # We reverse sort order at the end so later dates rank higher.
            date_rank = datetime.datetime.strptime(
                convert_date_string(
                    analysis['date_created']
                ),
                "%Y-%m-%dT%H:%M:%S.%f%z"
            )
            # Store all the ranking numbers for an analysis in a tuple that can be sorted.
            analysis_ranking[
                (
                    status_rank,
                    lab_rank,
                    pipeline_award_rfa_rank,
                    assembly_rank,
                    genome_annotation_rank,
                    pipeline_version_rank,
                    date_rank,
                )
            ] = analysis['@id']
        if analysis_ranking:
            # We want highest version, date, etc. so reverse sort order. Access value from top item.
            return sorted(analysis_ranking.items(), reverse=True)[0][1]
```

Reply:
refactored